modules/lang-painless/src/test/java/org/elasticsearch/painless/ElvisTests.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless;
+
+import static java.util.Collections.singletonMap;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.lessThan;
+
+/**
+ * Tests for the Elvis operator ({@code ?:}).
+ */
+public class ElvisTests extends ScriptTestCase {
+    public void testBasics() {
```

Comment:
Painless is neither Groovy nor Java.  The only reason the boxed types are even whitelisted is so that other whitelisted methods that require the boxed types can be used.  One bonus is that we allow unboxed types to call boxed type methods.  But otherwise, there is no reason I can think of to use a boxed type.  Eliminating boxing support for these types other than that situations I listed above has removed immense amounts of complexity from both casting and promotion.  I strongly stand by this decision and will continue to do so.

On the boxing situation for the elvis operator I would make it work exactly as the regular conditional does right now.  If a regular conditional had ```Integer a = Integer.valueOf(0); int b = a != null ? a : 1;``` I would expect it to throw a CCE.  Divergent behavior between the two operators will be even more confusing, not to mention the odds of someone actually hitting this is pretty low considering the def type should still work.  ```def a = Integer.valueOf(0); int b = a ?: 1;``` should work without any issues.